File: frontend/src/services/resourceService.js
/**
 * ================================================================
 * resourceService.js — Facilities & Assets API Layer
 * ================================================================
 * Owner: Member 3 - Facilities & Resources Module
 *
 * All functions return the response data directly (axios unwrapped).
 * Errors propagate to the caller for toast/alert handling.
 *
 * BASE: http://localhost:8081/api/resources
 *
 * FILTERING:
 *   getResources(filters) builds a query string from the filters
 *   object. Only non-empty keys are included so we never send
 *   ?type=&capacity= (which the backend would reject or ignore).
 *   Example: getResources({ type: 'ROOM', capacity: 50 })
 *     → GET /api/resources?type=ROOM&capacity=50
 * ================================================================
 */
import api from './axiosConfig'

const BASE = '/api/resources'

export const resourceService = {

  /**
   * GET /api/resources  — with optional filters
   * @param {Object} filters  { type, capacity, location, status }
   *   All keys are optional. Omit or pass null/empty to skip.
   */
  getResources: (filters = {}) => {
    // Build query params — skip null / empty-string values
    const params = {}
    if (filters.type)             params.type     = filters.type
    if (filters.capacity)         params.capacity = filters.capacity
    if (filters.location?.trim()) params.location = filters.location.trim()
    if (filters.status)           params.status   = filters.status

    return api.get(BASE, { params }).then(r => r.data)
  },

  /** GET /api/resources/available — resources where isAvailable=true */
  getAvailable: () => api.get(`${BASE}/available`).then(r => r.data),

  /** GET /api/resources/:id */
  getById: (id) => api.get(`${BASE}/${id}`).then(r => r.data),

  /**
   * POST /api/resources  — ADMIN only
   * @param {Object} data  ResourceRequestDTO shape
   */
  createResource: (data) => api.post(BASE, data).then(r => r.data),

  /**
   * PUT /api/resources/:id  — ADMIN only
   * @param {string} id   MongoDB resource id
   * @param {Object} data ResourceRequestDTO shape
   */
  updateResource: (id, data) => api.put(`${BASE}/${id}`, data).then(r => r.data),

  /**
   * DELETE /api/resources/:id  — ADMIN only
   * Returns void (204 No Content from backend).
   */
  deleteResource: (id) => api.delete(`${BASE}/${id}`),
}

File: backend/src/main/java/com/smartcampus/service/ResourceService.java
package com.smartcampus.service;

import com.smartcampus.dto.ResourceRequestDTO;
import com.smartcampus.dto.ResourceResponseDTO;
import com.smartcampus.model.ResourceStatus;
import com.smartcampus.model.ResourceType;

import java.util.List;

/**
 * ================================================================
 * ResourceService — Interface for Facilities & Assets Module
 * ================================================================
 * Defines the business-logic contract for resource management.
 * Implemented by ResourceServiceImpl.
 *
 * ROLE RULES (enforced in controller via @PreAuthorize):
 *   ADMIN → createResource, updateResource, deleteResource
 *   USER  → getAllResources, getResourceById, filterResources
 *
 * FILTERING:
 *   filterResources() accepts nullable parameters. When a param is
 *   null it is treated as "no filter on that field", allowing any
 *   combination of type / location / capacity / status to be used.
 * ================================================================
 */
public interface ResourceService {

    /**
     * Create a new campus resource.
     * Called by: POST /api/resources (ADMIN only)
     *
     * @param request validated inbound DTO
     * @return the persisted resource as a response DTO
     */
    ResourceResponseDTO createResource(ResourceRequestDTO request);

    /**
     * Retrieve every resource (no filter).
     * Called by: GET /api/resources (with no query params)
     *
     * @return unfiltered list of all resources
     */
    List<ResourceResponseDTO> getAllResources();

    /**
     * Retrieve a single resource by its MongoDB ObjectId.
     * Called by: GET /api/resources/{id}
     *
     * @param id MongoDB document id
     * @return resource DTO if found, throws ResourceNotFoundException otherwise
     */
    ResourceResponseDTO getResourceById(String id);

    /**
     * Full update (replace) an existing resource.
     * Called by: PUT /api/resources/{id} (ADMIN only)
     *
     * @param id      MongoDB document id
     * @param request validated update payload
     * @return updated resource DTO
     */
    ResourceResponseDTO updateResource(String id, ResourceRequestDTO request);

    /**
     * Permanently delete a resource.
     * Called by: DELETE /api/resources/{id} (ADMIN only)
     *
     * @param id MongoDB document id
     */
    void deleteResource(String id);

    /**
     * Multi-parameter filtering — the main search endpoint.
     * Called by: GET /api/resources?type=&capacity=&location=&status=
     *
     * All parameters are optional (nullable). The service applies
     * only the filters that are non-null, enabling any combination:
     *   ?type=ROOM                        → type filter only
     *   ?type=LAB&capacity=30             → type + capacity
     *   ?location=block+a&status=ACTIVE   → location + status
     *   ?type=ROOM&capacity=50&status=ACTIVE → all three + status
     *
     * @param type        filter by ResourceType enum (nullable)
     * @param minCapacity filter by minimum capacity (nullable)
     * @param location    partial/case-insensitive location match (nullable)
     * @param status      filter by ResourceStatus enum (nullable)
     * @return filtered list (empty list if nothing matches)
     */
    List<ResourceResponseDTO> filterResources(
            ResourceType type,
            Integer minCapacity,
            String location,
            ResourceStatus status
    );

    /**
     * Convenience method — return only resources marked as available.
     * Kept for backward-compatibility with BookingService callers.
     * Called by: GET /api/resources/available
     *
     * @return list of resources where isAvailable == true
     */
    List<ResourceResponseDTO> getAvailableResources();
}
